File: backtracking/Subset_Sum.cpp
#include <bits/stdc++.h>
using namespace std;

int arr[20];
int cnt = 0;
int n, s;


void solve(int k, int result){
    if(k==n){
        if(result==s) cnt++;
        return;
    }
    
    solve(k+1, result);
    solve(k+1, result+arr[k]);
}

int main(void){
    ios::sync_with_stdio(0);
    cin.tie(0);
    cin>>n>>s;
    for(int i=0;i<n;i++) cin>>arr[i];
    solve(0,0);
    if(s==0) cnt--;
    cout<<cnt;
}

File: backtracking/Subset_Sum.py
import sys
input = sys.stdin.readline

cnt = 0

def solve(k,n,s,result,arr):
    global cnt
    if k==n:
        if result==s: cnt += 1
        return
    
    solve(k+1,n,s,result,arr)
    solve(k+1,n,s,result+arr[k],arr)


def main():
    global cnt
    n,s = map(int, input().split())
    arr = list(map(int, input().split()))
    solve(0,n,s,0,arr)
    if s==0: cnt -= 1
    print(cnt)

if __name__ == "__main__":
    main()
